lib/select.ts

```diff
@@ -0,0 +1,201 @@
+ *    select(fruit, () => ["apple", "banana", "mango"]);
+ *
+ *    const employee = observable(17);
+ *    const employeesCB = () => [
+ *      {value: 12, label: "Bob", disabled: true},
+ *      {value: 17, label: "Alice"},
+ *      {value: 21, label: "Eve"},
+ *    ];
+ *    select(employee, employeesCB, {defaultLabel: "Select employee:"});
+ */
+export function select<T>(
+  obs: Observable<T>,
+  callback: () => Array<IOption<T>>,
+  options: ISelectUserOptions = {},
+  renderOption: (option: IOptionFull<T|null>) => DomArg = (option) => option.label
+): Element {
+  // Create SelectKeyState to manage user value search inputs.
+  const keyState = new SelectKeyState<T>(callback);
+
+  // Computed contains the IOptionFull of the obs value.
+  const selected = Computed.create(null, obs, (use, val) => {
+    const option = callback().find(_op => val === getOptionFull(_op).value);
+    return option ? getOptionFull(option) : ({value: null, label: ''} as IOptionFull<null>);
+  });
+
+  const container: Element = cssSelectBtn({tabIndex: '0', class: options.buttonCssClass || ''},
+    dom.autoDispose(selected),
+    dom.domComputed(selected, sel =>
+      renderOption(sel.label ? sel : {value: null, label: options.defaultLabel || ""})),
+    options.buttonArrow,
+    dom.on('keydown', (ev) => {
+      const sel = keyState.add(ev.key);
+      if (sel) { obs.set(sel.value); }
+    }),
+    (elem) => baseElem((...args) => Select.create(...args), elem, () => [
+      (elem) => stretchMenuToContainer(elem, container),
+      dom.forEach(callback(), (option) => {
+        const obj: IOptionFull<T> = getOptionFull(option);
+        // Note we only set 'selected' when an <option> is created; we are not subscribing to obs.
+        // This is to reduce the amount of subscriptions, esp. when number of options is large.
+        return menuItem(() => { obs.set(obj.value); },
+          Object.assign({ disabled: obj.disabled, selected: obj.value === obs.get() },
+            obj.disabled ? { class: 'disabled'} : {}),
+          renderOption(obj)
+        );
+      }),
+    ], {
+      menuCssClass: options.menuCssClass,
+      trigger: [(triggerElem: Element, ctl: PopupControl) => {
+        dom.onElem(triggerElem, 'click', () => ctl.toggle()),
+        dom.onKeyElem(triggerElem as HTMLElement, 'keydown', {
+          ArrowDown: () => ctl.open(),
+          ArrowUp: () => ctl.open()
+        })
+      }],
+      selectLabelOnOpen: () => selected.get().label
+    })
+  );
+  return container;
+}
+
+/**
+ * Creates an instance of Menu intended to mimic the behavior of the select HTML element.
+ *
+ * Should always be created using select(), which adds the select button and associated logic.
+ */
+class Select<T> extends BaseMenu {
+  private _selectRows: HTMLElement[] = Array.from(this.content.children) as HTMLElement[];
+  private _keyState: SelectKeyState<string> = new SelectKeyState(() =>
+    this._selectRows.map(_elem => ({
```

Comment:
Would it make sense to cache the list of options, so that the callback can just be `() => this._options`? (To reduce the work done on each keystroke?)